src/cmap/connect.ts

```diff
@@ -1,28 +1,50 @@
-import net = require('net');
-import tls = require('tls');
-import { Connection } from './connection';
+import * as net from 'net';
+import * as tls from 'tls';
+import { Connection, MongoDBConnectionOptions } from './connection';
 import { MongoError, MongoNetworkError, MongoNetworkTimeoutError } from '../error';
-import { defaultAuthProviders } from './auth/defaultAuthProviders';
+import { AUTH_PROVIDERS } from './auth/defaultAuthProviders';
 import { AuthContext } from './auth/auth_provider';
-import { makeClientMetadata } from '../utils';
+import { makeClientMetadata, ClientMetadataOptions, ClientMetadata } from '../utils';
 import {
   MAX_SUPPORTED_WIRE_VERSION,
   MAX_SUPPORTED_SERVER_VERSION,
   MIN_SUPPORTED_WIRE_VERSION,
   MIN_SUPPORTED_SERVER_VERSION
 } from './wire_protocol/constants';
-const AUTH_PROVIDERS: any = defaultAuthProviders();
-
-function connect(options: any, cancellationToken: any, callback: Function) {
-  if (typeof cancellationToken === 'function') {
-    callback = cancellationToken;
+import type { Callback, CallbackWithType, Document, UniversalError } from '../types';
+import type { EventEmitter } from 'events';
+
+import type { Socket, SocketConnectOpts } from 'net';
+import type { TLSSocket, ConnectionOptions as TLSConnectionOpts } from 'tls';
+
+type UniversalSocket = Socket | TLSSocket;
+
+export function connect(
+  options: MongoDBConnectionOptions,
+  callback: Callback<Connection | UniversalSocket>
+): void;
+export function connect(
+  options: MongoDBConnectionOptions,
+  cancellationToken: EventEmitter,
+  callback: Callback<Connection | UniversalSocket>
+): void;
+export function connect(
+  options: MongoDBConnectionOptions,
+  _cancellationToken: EventEmitter | Callback<Connection | UniversalSocket>,
+  _callback?: Callback<Connection | UniversalSocket>
+): void {
+  let cancellationToken = _cancellationToken as EventEmitter | undefined;
+  const callback = (_callback ?? _cancellationToken) as Callback<Connection | UniversalSocket>;
+  if ('function' === typeof cancellationToken) {
```

Comment:
I wanted the function signatures to have cancellationToken to be either required or omitted all together, it gives you the two distinct call signatures where you can drop the token. Here I cast to a union of EventEmitter | undefined because in the function if statement I overwrite the cancellationToken with undefined. It's all meant to resolve the complications with typing and function overloads by hiding it in the function instead of having weird signatures.

Although maybe the signatures are still not entirely accurate then here?

Reply:
I checked the emitted output and it looks fine, but it does look strange to me. It looks like you cast it to `EventEmitter | undefined` above, if that were the case this could never be a `function`. Is there a way to rewrite this so the validation is a little more obvious?